### ISSUE TITLE
[Bug]: retriever with similarity_top_k sometimes return no nodes at all

### ISSUE THREAD

**Issue by Calabrone76** [run-llama/llama_index#14302]
### Bug Description

We used a retriever with filters connected to a VectorStore (PGVector). It works, but sometimes with some specific sentence it doen't retrieve anything.

Here is the the method we used to create the retriever ad use it in a chat engine

```
  def setIndex(self, category: int):    
    openai.api_key = self.__llm.api_key
    db = vdb(self.config['database'])
    qry = f"""SELECT documents.id FROM documents WHERE documents.id IN (
              SELECT distinct documents.id from documents, catedoc
              WHERE documents.id = catedoc.id_doc AND
              catedoc.id_cat IN
              (
              WITH RECURSIVE ctename AS (
              SELECT categories.id
              FROM categories
              WHERE categories.id = {category}
              UNION ALL
              SELECT categories.id
              FROM categories JOIN ctename ON categories.refid = ctename.id
              ) SELECT * FROM ctename ) )"""        
    list_cat = [str(item['id']) for item in db.query(qry)]        
    v_store = db.db_vector_store('documents_view')
    self.index = VectorStoreIndex.from_vector_store(vector_store=v_store)    
    filters = MetadataFilters(
      filters=[
          MetadataFilter(key="id", value=list_cat, operator=FilterOperator.IN)                   
      ]      
    ) 
    retriever = self.index.as_retriever(
      filters=filters,
      similarity_top_k=20,      
      verbose=True,
    )            
    chat = ContextChatEngine(
        retriever=retriever, 
        llm=self.__llm, 
        memory=self.__memory, 
        prefix_messages=[]        
      ).from_defaults(
        retriever=retriever,         
        system_prompt=ConversationLLamaIndex.__MISSION    
      )
    
    self.lama_chat = chat
```

Here the method that extract the vectorStore:
```
    def db_vector_store(self, table_name: str) -> PGVectorStore :
        return PGVectorStore.from_params(
            database=self._config["dbname"],
            host=self._config["hostname"],
            password=self._config["password"],
            port=5432,
            user=self._config["username"],
            table_name=table_name,
            embed_dim=1536            
        )
```

We can't use index.as_chat_engine because in that case it's not possible tu set filters.

### Version

0.10.45

### Steps to Reproduce

Create a chat engine and pass a retriever as parameter as shown in the code above, than use the chat method to ask questions.

**Comment by logan-markewich** [comment]
Wouldn't that mean the filter is just filtering out all nodes? 

I also have no idea what `v_store = db.db_vector_store('documents_view')` is actually doing or returning

**Comment by schiaro98** [comment]
Hi, and thank you for the response Logan. The db_vector_store returns the postgres vector_store that contains all the embeddings. The filters are supposed to work because all other query that uses the same filters returns nodes. 
We would like to know how the cosine similarity is configured, the suspicion is that results with too low scores are not returned, as demonstrated by the fact that similar queries in the text are answer correctly, and nodes are reported

**Comment by Calabrone76** [comment]
> Wouldn't that mean the filter is just filtering out all nodes?
> 
> I also have no idea what `v_store = db.db_vector_store('documents_view')` is actually doing or returning. This is hardly a reproducible piece of code 😅

Hi and thank's for the quick reply. As my collegue @schiaro98 said, we had stored all embeddings in a PGVector db and we extract them with that method. Each embedding is marked with a category_id that we filter to get the right category on which perform retrieval.

**Comment by logan-markewich** [comment]
@Calabrone76 @schiaro98 still a little confused I guess. There shouldn't be any score filtering. The only way that no nodes are returned is if the filters remove all result, or if the db actually has no data.

You can try testing the retriever directly

`nodes = retriever.retrieve("test")`

**Comment by Calabrone76** [comment]
> The only way that no nodes are returned is if the filters remove all result, or if the db actually has no data.

Hi @logan-markewich, the problem is that the retriever and his filters are created only one time when the user choose a category. When the user asks question on the same category the retriever remain the same: the odd thing is that with some question no nodes are returned; if we refactor the question then nodes are returned. 
We can reproduce the issue even testing the retriever directly.

**Comment by Calabrone76** [comment]
HI @logan-markewich, we made some more debug and we noticed that every time we use the chat method it always issues query to our PGvector DB. We thought that once the retriever had been created with filters all the data was on the retriever itself. But it seems, and it makes sense, the every similarity search is done against the DB itself. So, if this is the case, maybe the problem is on the PGVector class.
